src/vectrick.jl

```diff
@@ -47,44 +40,197 @@ function mul!(x::AbstractVector, K::AbstractKroneckerProduct, v::AbstractVector)
     return x
 end
 
-function Base.:*(K::AbstractKroneckerProduct, v::AbstractVector)
-    return mul!(Vector{promote_type(eltype(v), eltype(K))}(undef, first(size(K))), K, v)
+
+if VERSION < v"1.3.0-alpha.115"
+    function mul_vec_trick!(x::AbstractVector, K::AbstractKroneckerSum, v::AbstractVector)
+        A, B = getmatrices(K)
+        a, b = size(A)
+        c, d = size(B)
+        e = length(v)
+        f = length(x)
+
+        V = reshape(v, d, b)
+        X = reshape(x, c, a)
+        mul!(X, V, transpose(A))
+        X .+= B * V
+        return x
+    end
+else # 5-arg mul! is available
+    function mul_vec_trick!(x::AbstractVector, K::AbstractKroneckerSum, v::AbstractVector)
+        A, B = getmatrices(K)
+        a, b = size(A)
+        c, d = size(B)
+        e = length(v)
+        f = length(x)
+
+        V = reshape(v, d, b)
+        X = reshape(x, c, a)
+        mul!(X, V, transpose(A))
+        mul!(X, B, V, true, true)
+        return x
+    end
+end # VERSION
+
+
+function mul_vec_trick!(X::AbstractMatrix, A::GeneralizedKroneckerProduct, V::AbstractMatrix)
+    @inbounds for i in eachindex(axes(X, 2), axes(V, 2))
+        @views mul_vec_trick!(X[:, i], A, V[:, i])
+    end
+    return X
 end
 
-reshape_cols(x, sizes...) = reshape(x, sizes...)
-reshape_rows(x, sizes...) = transpose(reshape(collect(transpose(x)), reverse(sizes)...))
 
-kron_id_a(a, x) = reshape_cols(a * reshape_cols(x, size(a)[2], :), :, size(x)[2])
-kron_a_id(a, x) = reshape_rows(a * reshape_rows(x, size(a)[2], :), :, size(x)[2])
```

Comment:
Main reason i just outright deleted those was because they weren't exported at all, so i'm not sure if it could be considered a breaking change

If I understand correctly, these methods were written to "skip" identity matrices when applying the vectrick recursively. Not sure if they're needed anymore given that vectrick is now only used in the case of 2 kronecker factors. For this reason I didn't compare the runtimes.

Now that you mention it, maybe these methods could give a slight speed boost in the 2-factor case, but I'll have to leave that for later.